### PR TITLE
GoogleAuthenticatorAccount set a unique id

### PR DESCRIPTION
**Pull request by Al1K4zu**
Hi,
first of all, thank you for this application.
When using MFA with Google Authenticator, the ID generated was the same for all users. So I modified GoogleAuthenticatorAccount.java to give a unique id to each user.
I simply use java.lang.System.currentTimeMillis() as unique id.
Thank you for merging my modification.

### CODE REVIEW

**Comment by mmoayyed** [comment]
Thanks for the change.

Could you please explain why you see ids being the same? Have you looked into the cause to figure out why this happens?

**Comment by Al1K4zu** [comment]
In OneTimeTokenAccount, the id is set to Integer.MAX_VALUE (2147483647) and is never changed.
I use mongodb repository and each time a new user is coming, cas replaces the existing one in mongodb collection.
example
first user = mongoReplCas:PRIMARY> db["MongoDbGoogleAuthenticatorRepository"].find()
{ "_id" : NumberLong(2147483647), "_class" : "org.apereo.cas.adaptors.gauth.repository.credentials.GoogleAuthenticatorAccount", "secretKey" : "UQYMMXRE7ZJFTFBJ", "validationCode" : 816033, "scratchCodes" : [ 27147845, 47267547, 73735658, 53856834, 83711853 ], "username" : "user1" }
second user = mongoReplCas:PRIMARY> db["MongoDbGoogleAuthenticatorRepository"].find()
{ "_id" : NumberLong(2147483647), "_class" : "org.apereo.cas.adaptors.gauth.repository.credentials.GoogleAuthenticatorAccount", "secretKey" : "MVRJ2ZHXAOB6MWLQ", "validationCode" : 802999, "scratchCodes" : [ 21196666, 99519535, 63367006, 99220901, 93312634 ], "username" : "user2" }

with my modification, the id is the timestamp
example :
{ "_id" : NumberLong("1504536320781"), "_class" : "org.apereo.cas.adaptors.gauth.repository.credentials.GoogleAuthenticatorAccount", "secretKey" : "IABQ5EO3EZ4BEPZ7", "validationCode" : 117895, "scratchCodes" : [ 96391108, 52454688, 22294482, 71170984, 57882634 ], "username" : "user1" }

**Comment by mmoayyed** [comment]
Thank you for the explanation. This makes very good sense.

A few things:

1. Please fix the build failures; usually this has to do with formatting, etc.
2. I recommend that you add a few test cases to ensure this bug is correctly fixed and verified.
3. When you're done, please post a similar PR for master and we'd be happy to merge both.

Thanks again.

**Comment by Al1K4zu** [comment]
I don't know what to change because the error message is  not explicit.
I think the probleme is not in my modification.
What I see in the log file : 
[ant:checkstyle] [INFO] /home/travis/build/apereo/cas/core/cas-server-core-tickets/src/main/java/org/apereo/cas/ticket/ServiceTicketImpl.java:123:5: Definition of 'equals()' without corresponding definition of 'hashCode()'. [EqualsHashCode]

I am not familiar with git so I do what I can.

**Comment by mmoayyed** [comment]
That has nothing to do with this issue. It's just INFO. INFO means information. You can review the build failures here on the PR request noted by `continuous-integration/travis-ci/pr`, examine the log and make fixes. You see how it says "Some checks were not successful" ? Click details and examine the log.

**Comment by mmoayyed** [comment]
And you see the following, if I may:

```
[ant:checkstyle] [ERROR] /home/travis/build/apereo/cas/support/cas-server-support-gauth-core/src/main/java/org/apereo/cas/adaptors/gauth/repository/credentials/GoogleAuthenticatorAccount.java:41:14: Variable 'v' should be declared final. [FinalLocalVariable]
[ant:checkstyle] [ERROR] /home/travis/build/apereo/cas/support/cas-server-support-gauth-core/src/main/java/org/apereo/cas/adaptors/gauth/repository/credentials/GoogleAuthenticatorAccount.java:41:18: Use a single space to separate non-whitespace characters. [SingleSpaceSeparator]
```

File: support/cas-server-support-gauth-core/src/main/java/org/apereo/cas/adaptors/gauth/repository/credentials/GoogleAuthenticatorAccount.java [hunk 1]
```diff
@@ -38,5 +38,7 @@ public GoogleAuthenticatorAccount(@JsonProperty("username") final String usernam
                                       @JsonProperty("validationCode") final int validationCode,
                                       @JsonProperty("scratchCodes") final List<Integer> scratchCodes) {
         super(username, secretKey, validationCode, scratchCodes);
+        long v =  java.lang.System.currentTimeMillis();
```
Comment (mmoayyed) [inline support/cas-server-support-gauth-core/src/main/java/org/apereo/cas/adaptors/gauth/repository/credentials/GoogleAuthenticatorAccount.java]: Also `v` is probably not a very good variable name. Either choose `id` or something more apparent, or remove the variable and set directly.

**Comment by Al1K4zu** [comment]
I removed the variable v.
There's a new compilation error, not due by my modification : 
org.apereo.cas.authentication.MongoAuthenticationHandlerTests > verifyAuthentication FAILED
    javax.security.auth.login.FailedLoginException: Failed to validate credentials: No account found for: u1

:support:cas-server-support-mongo:test FAILED

**Comment by mmoayyed** [comment]
Excellent. Merge with upstream once, and then issue should go away. Then start working on the other items.

**Comment by Al1K4zu** [comment]
all checks have passed
Can you merge ?
thank you

**Comment by Al1K4zu** [comment]
I don't understand why it's so complicated to integrate a simple bug fix.
I will not do test cases or any other task for this pull request.
You did not create those test cases by yourself when you created this method and, now,  you ask me to do it.
For a simple code line,  is it normal to spend so much time ?
I corrected some others bugs (and there are plenty) but I will not do pull requests anymore.
Bye.

**Comment by mmoayyed** [comment]
Not a problem. Thanks for your participation. Whenever you feel inclined to resume work please post back and we can continue. Thanks again.

**Comment by puntakana** [comment]
I think Al1K4zu corrects this MAJOR bug https://groups.google.com/a/apereo.org/forum/?utm_medium=email&utm_source=footer#!msg/cas-user/vDEfJ6SxlwY/I5pTug9ODQAJ

How can I help to add this test case because I test his patch and it's OK with it !!

**Comment by mmoayyed** [comment]
Very good. What you want to do is provide a PR that is targetted against 5.1, along with a set of test cases prefererably where you test the patch against `MongoDbGoogleAuthenticatorTokenCredentialRepository`. So you want to write a `MongoDbGoogleAuthenticatorTokenCredentialRepositoryTests` similiar to `MongoDbGoogleAuthenticatorAccountRegistryTests` and ensure your changes are actually correct. 

When that's all good and done, you are welcome to post the same PR targeted against master or ask me and I can port it forward for you.

**Comment by mmoayyed** [comment]
And more importantly, all you need to do is add `@org.springframework.data.annotation.Id` to the id field. That's all.

**Pull request closed without merging**